server/src/main/java/org/elasticsearch/transport/SniffConnectionStrategy.java

```diff
@@ -0,0 +1,273 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.transport;
+
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.apache.lucene.util.SetOnce;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.StepListener;
+import org.elasticsearch.action.admin.cluster.state.ClusterStateAction;
+import org.elasticsearch.action.admin.cluster.state.ClusterStateRequest;
+import org.elasticsearch.action.admin.cluster.state.ClusterStateResponse;
+import org.elasticsearch.cluster.ClusterName;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.common.collect.Tuple;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.transport.TransportAddress;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+import org.elasticsearch.core.internal.io.IOUtils;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.util.Iterator;
+import java.util.List;
+import java.util.function.Consumer;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+
+public class SniffConnectionStrategy extends RemoteConnectionStrategy {
+
+    private final String clusterAlias;
+    private final List<Tuple<String, Supplier<DiscoveryNode>>> seedNodes;
+    private final TransportService transportService;
+    private final int maxNumRemoteConnections;
+    private final Predicate<DiscoveryNode> nodePredicate;
+    private final SetOnce<ClusterName> remoteClusterName = new SetOnce<>();
+    private volatile String proxyAddress;
+
+    SniffConnectionStrategy(String clusterAlias, TransportService transportService, RemoteConnectionManager connectionManager,
+                            String proxyAddress, int maxNumRemoteConnections, Predicate<DiscoveryNode> nodePredicate,
+                            List<Tuple<String, Supplier<DiscoveryNode>>> seedNodes) {
+        super(transportService.getThreadPool(), connectionManager);
+        this.clusterAlias = clusterAlias;
+        this.transportService = transportService;
+        this.proxyAddress = proxyAddress;
+        this.maxNumRemoteConnections = maxNumRemoteConnections;
+        this.nodePredicate = nodePredicate;
+        this.seedNodes = seedNodes;
+    }
+
+    @Override
+    protected void connectImpl(ActionListener<Void> listener) {
+        collectRemoteNodes(seedNodes.stream().map(Tuple::v2).iterator(), listener);
+    }
+
+    @Override
+    protected boolean shouldOpenMoreConnections() {
+        return connectionManager.size() < maxNumRemoteConnections;
+    }
+
+    private void collectRemoteNodes(Iterator<Supplier<DiscoveryNode>> seedNodes, ActionListener<Void> listener) {
+        if (Thread.currentThread().isInterrupted()) {
+            listener.onFailure(new InterruptedException("remote connect thread got interrupted"));
```

Comment:
I know this was already in the original code, but we should probably return here after calling the listener